src/ScalarMessaging.js

```diff
@@ -0,0 +1,171 @@
+        url = url.substr(0, url.length - 1);
+    }
+    if (url !== event.origin) {
+        console.warn("Unauthorised postMessage received. Source URL: " + event.origin);
+        return;
+    }
+
+    switch (event.data.action) {
+        case "membership_state": {
+            const roomId = event.data.room_id;
+            const userId = event.data.user_id;
+            if (!userId) {
+                return sendError(event, "Missing user_id in request");
+            }
+            if (!roomId) {
+                return sendError(event, "Missing room_id in request");
+            }
+            console.log(`membership_state of ${userId} in room ${roomId} requested.`);
+            const client = MatrixClientPeg.get();
+            if (!client) {
+                return sendError(event, "You need to be logged in.");
+            }
+            const room = client.getRoom(roomId);
+            if (!room) {
+                return sendError(event, "This room is not recognised.");
+            }
+            let membershipState = "leave";
+            const member = room.getMember(userId);
+            if (member) {
+                membershipState = member.membership;
+            }
+            sendResponse(event, {
+                membership_state: membershipState
+            });
+        }
+            break;
+        case "invite": {
+            const roomId = event.data.room_id;
+            const userId = event.data.user_id;
+            if (!userId) {
+                return sendError(event, "Missing user_id in request");
+            }
+            if (!roomId) {
+                return sendError(event, "Missing room_id in request");
+            }
+            console.log(`Received request to invite ${userId} into room ${roomId}`);
+            const client = MatrixClientPeg.get();
+            if (!client) {
+                return sendError(event, "You need to be logged in.");
+            }
+            const room = client.getRoom(roomId);
+            if (room) {
+                // if they are already invited we can resolve immediately.
+                const member = room.getMember(userId);
+                if (member && member.membership === "invite") {
+                    sendResponse(event, {
+                        invite: true
+                    });
+                    return;
+                }
+            }
+
+            client.invite(roomId, userId).then(function() {
+                sendResponse(event, {
+                    invite: true
+                });
+            }, function(err) {
+                sendError(event, "You need to be able to invite users to do that.", err);
+            });
+        }
+            break;
+        default:
+            console.warn("Unhandled postMessage event with action '" + event.data.action +"'");
+            break;
+    }
+};
+
+function sendResponse(event, res) {
+    console.log("Action:" + event.data.action + " succeeded with response: " + JSON.stringify(res));
```

Comment:
I don't know how spammy/not spammy we are expected to be in vector-land. Logging points like this can be useful for debugging obscure failure modes, but if the general rule is to be quiet I can remove these.